File: src/broadcast-box.ts
import ky, { type KyInstance } from "ky";
import { z } from "zod";
import { logger } from "./logger.ts";

/**
 * Lenient schema for a Broadcast Box `StreamSessionState` (the admin status
 * endpoint returns a JSON array of these). Only the fields we rely on are
 * described; any others are preserved by `.loose()`.
 */
const streamSessionSchema = z
  .object({
    streamKey: z.string().optional(),
    videoTracks: z.array(z.unknown()).optional(),
    audioTracks: z.array(z.unknown()).optional(),
  })
  .loose();

/**
 * The admin status endpoint returns a JSON array of sessions, or `null` when
 * no streams exist at all. Normalise the `null` case to an empty array.
 */
const statusSchema = z
  .array(streamSessionSchema)
  .nullable()
  .transform((sessions) => sessions ?? []);

type StreamSession = z.infer<typeof streamSessionSchema>;

function hasEntries(value: readonly unknown[] | undefined): boolean {
  return value !== undefined && value.length > 0;
}

/**
 * A stream counts as live when it exposes a stream key and is actively
 * publishing — signalled by any received media track. `streamStart` is set as
 * soon as a stream key is provisioned (even while offline), so it can't be
 * used to decide liveness.
 */
function isLive(state: StreamSession): state is StreamSession & { streamKey: string } {
  return (
    typeof state.streamKey === "string" &&
    state.streamKey.trim() !== "" &&
    (hasEntries(state.videoTracks) || hasEntries(state.audioTracks))
  );
}

export class BroadcastBoxClient {
  readonly #api: KyInstance;

  constructor(options: { apiUrl: string; authorization: string }) {
    this.#api = ky.create({
      prefix: options.apiUrl,
      headers: {
        Authorization: options.authorization,
        Accept: "application/json",
      },
      retry: { limit: 2 },
      timeout: 10_000,
    });
  }

  /**
   * Fetches the currently live stream keys from `/api/admin/status`.
   *
   * The admin endpoint is used because Broadcast Box runs with
   * `DISABLE_STATUS=true`, which turns off the public `/api/status` route.
   *
   * Throws `ky.HTTPError` on a non-2xx response and `SchemaValidationError`
   * when the payload does not match the expected shape.
   */
  async fetchLiveStreamKeys(signal?: AbortSignal): Promise<Set<string>> {
    const sessions = await this.#api.get("api/admin/status", { signal }).json(statusSchema);

    const live = new Set<string>();

    for (const session of sessions) {
      if (isLive(session)) {
        live.add(session.streamKey);
      }
    }

    logger.debug(`Broadcast Box reports ${live.size} live stream(s)`);

    return live;
  }
}

File: internal/broadcastbox/broadcastbox.go
// Package broadcastbox reads the live stream keys from a Broadcast Box server's
// admin status endpoint.
package broadcastbox

import (
	"context"
	"encoding/json"
	"fmt"
	"strings"
	"time"

	"github.com/go-resty/resty/v2"

	"github.com/timche/teamspeak-stream-live/internal/httpx"
	"github.com/timche/teamspeak-stream-live/internal/logger"
)

const requestTimeout = 10 * time.Second

// streamSession is a lenient view of a Broadcast Box `StreamSessionState`. Only
// the fields we rely on are described; unknown fields are ignored by the decoder.
type streamSession struct {
	StreamKey   string `json:"streamKey"`
	VideoTracks []any  `json:"videoTracks"`
	AudioTracks []any  `json:"audioTracks"`
}

// isLive reports whether a session exposes a stream key and is actively
// publishing (signalled by any received media track). `streamStart` is set as
// soon as a key is provisioned, so it cannot be used to decide liveness.
func isLive(s streamSession) bool {
	return strings.TrimSpace(s.StreamKey) != "" &&
		(len(s.VideoTracks) > 0 || len(s.AudioTracks) > 0)
}

// Options configures a Client.
type Options struct {
	APIURL string
	// Authorization is the ready-to-send header value, "Bearer <base64(token)>".
	Authorization string
}

// Client fetches live stream keys from Broadcast Box.
type Client struct {
	http *resty.Client
}

// New builds a Client for the given Broadcast Box endpoint.
func New(opts Options) *Client {
	c := resty.New().
		SetBaseURL(opts.APIURL).
		SetHeader("Authorization", opts.Authorization).
		SetHeader("Accept", "application/json").
		SetTimeout(requestTimeout).
		SetRetryCount(2)
	httpx.RetryOnStatuses(c)
	return &Client{http: c}
}

// FetchLiveStreamKeys fetches the currently live stream keys from
// `/api/admin/status`. The admin endpoint is used because Broadcast Box runs
// with DISABLE_STATUS=true. A `null` body (no streams) is treated as empty.
func (c *Client) FetchLiveStreamKeys(ctx context.Context) (map[string]struct{}, error) {
	resp, err := c.http.R().SetContext(ctx).Get("api/admin/status")
	if err != nil {
		return nil, err
	}
	if resp.IsError() {
		return nil, fmt.Errorf("broadcast box status request failed: %d", resp.StatusCode())
	}

	var sessions []streamSession
	if len(resp.Body()) > 0 {
		if err := json.Unmarshal(resp.Body(), &sessions); err != nil {
			return nil, err
		}
	}

	live := make(map[string]struct{})
	for _, session := range sessions {
		if isLive(session) {
			live[session.StreamKey] = struct{}{}
		}
	}
	logger.Log.Debug("Broadcast Box live streams", "count", len(live))
	return live, nil
}
